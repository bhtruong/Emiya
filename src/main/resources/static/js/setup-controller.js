import { CARD_SETS, SETUP, CLEANUP } from './tests/data.js';

function getStubbedCardSets() {
    return new Promise(resolve => {
        setTimeout(() => {
            resolve(CARD_SETS);
        }, 1250);
    })
}

function getStubbedSetup(cardSets, players) {
    return new Promise(resolve => {
        setTimeout(() => {
            resolve(SETUP)
        }, 2500, cardSets, players)
    })
}

function getStubbedCleanup(cardSets) {
    return new Promise(resolve => {
        setTimeout(() => {
            resolve(CLEANUP)
        }, 3000, cardSets)
    })
}

const getRandomSetup = (cardSets, players) => {
    const url = `/api/randomSetup?players=${players}`;
    const options = {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(cardSets)
    };

    return fetch(url, options).then(response => {
        return response.json()
    })
};

export default {
    getCardSets: getStubbedCardSets,
    getSetup: getStubbedSetup,
    getCleanup: getStubbedCleanup
};
