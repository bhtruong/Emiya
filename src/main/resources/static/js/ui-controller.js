import { TYPE } from './tests/data.js';

const BUTTONS_PER_ROW = 4;

function renderCardSets(cardSets) {
    let cardSetsNode = document.querySelector('.cardSets');
    let rows = Math.ceil(cardSets.length / BUTTONS_PER_ROW);
    let rowElements = [];

    for (let i = 0; i < rows; i++) {
        let section = document.createElement('section');

        section.classList.add('row');

        rowElements.push(section)
    }

    rowElements.forEach((row, rowNumber) => {
        row.classList.add(`button-row-${rowNumber}`);
        cardSetsNode.appendChild(row)
    });

    cardSets.forEach((cardSet, index) => {
        let rowNumber = Math.floor(index / BUTTONS_PER_ROW);
        let row = document.querySelector(`.button-row-${rowNumber}`);

        let button = document.createElement('button');
        button.classList.add('btn-sm', 'btn-dark');
        button.setAttribute('type', 'button');
        button.innerText = cardSet;

        row.appendChild(button)
    });

    return cardSetsNode.innerHTML
}

function renderSingle(data) {
    let containerSelector = '', nameSelector = '';
    let image = document.createElement('img');

    image.setAttribute('src', data.imageURL);
    image.classList.add('card-image');

    if (data.type === TYPE.SCHEME) {
        containerSelector = '.scheme-container';
        nameSelector = '.scheme';
    } else if (data.type === TYPE.MASTERMIND) {
        containerSelector = '.mastermind-container';
        nameSelector = '.mastermind';
    }

    document.querySelector(containerSelector).appendChild(image);
    document.querySelector(nameSelector).textContent = data.name;
}

function renderMultiple(data) {
    let rowSelector = '', row = null;

    if (data.type === TYPE.VILLAINS) {
        rowSelector = '.villains'
    } else if (data.type === TYPE.HEROES) {
        rowSelector = '.heroes'
    }

    row = document.querySelector(rowSelector);

    data.elements.forEach(element => {
        let col = document.createElement('div');
        let name = document.createElement('p');
        let image = document.createElement('img');

        col.classList.add('col-sm');

        if (data.type === TYPE.VILLAINS) {
            name.classList.add('text-danger')
        }

        name.classList.add('image-label');
        name.innerHTML = element.name;

        image.setAttribute('src', element.imageURL);

        col.append(name);
        col.append(image);

        row.appendChild(col);
    })
}

function toggleButtonColor(button) {
    if (button.classList.contains('btn-dark')) {
        button.classList.replace('btn-dark', 'btn-light');
    } else if (button.classList.contains('btn-light')) {
        button.classList.replace('btn-light', 'btn-dark');
    }

    return button.classList
}

function renderSetup(setup) {
    renderSingle(setup['scheme']);
    renderSingle(setup['mastermind']);
    renderMultiple(setup['villains']);
    renderMultiple(setup['heroes']);

    return document.querySelector('.randomSetup').innerHTML
}

function renderCleanup(cleanup) {
    document.querySelector('.bystanders').textContent = cleanup['bystanders'];
    document.querySelector('.wounds').textContent = cleanup['wounds'];
    document.querySelector('.bindings').textContent = cleanup['bindings'];
    document.querySelector('.shards').textContent = cleanup['shards'];
    document.querySelector('.sidekicks').textContent = cleanup['sidekicks'];
    document.querySelector('.ambitions').textContent = cleanup['ambitions'];

    return document.querySelector('.cleanup-generator').innerHTML
}

function removeChildren(parent) {
    while (parent.lastChild) {
        parent.removeChild(parent.lastChild)
    }
}

function removeImage(parent) {
    let img = document.querySelector('.card-image');

    parent.removeChild(img)
}

function resetForm() {
    removeChildren(document.querySelector('.scheme'));
    removeChildren(document.querySelector('.mastermind'));
    removeChildren(document.querySelector('.villains'));
    removeChildren(document.querySelector('.heroes'));

    removeImage(document.querySelector('.scheme-container'));
    removeImage(document.querySelector('.mastermind-container'));

    return document.querySelector('.randomSetup').innerHTML
}

function resetCleanup() {
    removeChildren(document.querySelector('.bystanders'));
    removeChildren(document.querySelector('.wounds'));
    removeChildren(document.querySelector('.bindings'));
    removeChildren(document.querySelector('.shards'));
    removeChildren(document.querySelector('.sidekicks'));
    removeChildren(document.querySelector('.ambitions'));

    return document.querySelector('.cleanup-generator').innerHTML
}

export default {
    renderCardSets: renderCardSets,
    renderSetup: renderSetup,
    renderCleanup: renderCleanup,
    resetForm: resetForm,
    resetCleanup: resetCleanup,
    toggleButtonColor: toggleButtonColor
};
