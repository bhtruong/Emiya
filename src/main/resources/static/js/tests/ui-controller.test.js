import { CARD_SETS, SETUP, CLEANUP } from "./data.js";
import UIController from "../ui-controller.js";

beforeAll(() => {
    document.body.innerHTML = `
        <section class="cardSets"></section>
        
        <div class="randomSetup">
        
            <div class="scheme-container">
                <h4 class="text-danger">Scheme</h4>
                <p class="text-danger scheme"></p>
            </div>
            
            <div class="mastermind-container">
                <h4>Mastermind</h4>
                <p class="mastermind"></p>
            </div>
            
            <div class="villains-container">
                <h4 class="text-danger">Villains</h4>
                <ul class="villains row"></ul>
            </div>
            
            <div class="heroes-container">
                <h4>Heroes</h4>
                <ul class="heroes row"></ul>
            </div>
            
        </div>
        
        <div class="cleanup-generator">
            <h2>Clean Up</h2>
            <h4>Bystanders</h4>
            <p class="bystanders"></p>
            <h4>Wounds</h4>
            <p class="wounds"></p>
            <h4>Bindings</h4>
            <p class="bindings"></p>
            <h4>Shards</h4>
            <p class="shards"></p>
            <h4>Sidekicks</h4>
            <p class="sidekicks"></p>
            <h4>Ambitions</h4>
            <p class="ambitions"></p>
        </div>`;
});

describe("ui controller renders", () => {
    test("card sets correctly", () => {
        const html = UIController.renderCardSets(CARD_SETS);

        expect(html).toMatchSnapshot()
    });

    test("setup correctly", () => {
        const html = UIController.renderSetup(SETUP);

        expect(html).toMatchSnapshot()
    });

    test("cleanup correctly", () => {
        const html = UIController.renderCleanup(CLEANUP);

        expect(html).toMatchSnapshot()
    });

    test("form reset correctly", () => {
        const html = UIController.resetForm();

        expect(html).toMatchSnapshot()
    });

    test("cleanup reset correctly", () => {
        const html = UIController.resetCleanup();

        expect(html).toMatchSnapshot()
    });
});