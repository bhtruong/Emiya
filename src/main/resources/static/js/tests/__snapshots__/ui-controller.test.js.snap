// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui controller renders card sets correctly 1`] = `"<section class=\\"row button-row-0\\"><button class=\\"btn-sm btn-dark\\" type=\\"button\\"></button><button class=\\"btn-sm btn-dark\\" type=\\"button\\"></button><button class=\\"btn-sm btn-dark\\" type=\\"button\\"></button><button class=\\"btn-sm btn-dark\\" type=\\"button\\"></button></section><section class=\\"row button-row-1\\"><button class=\\"btn-sm btn-dark\\" type=\\"button\\"></button><button class=\\"btn-sm btn-dark\\" type=\\"button\\"></button><button class=\\"btn-sm btn-dark\\" type=\\"button\\"></button><button class=\\"btn-sm btn-dark\\" type=\\"button\\"></button></section><section class=\\"row button-row-2\\"><button class=\\"btn-sm btn-dark\\" type=\\"button\\"></button></section>"`;

exports[`ui controller renders cleanup correctly 1`] = `
"
            <h2>Clean Up</h2>
            <h4>Bystanders</h4>
            <p class=\\"bystanders\\">33</p>
            <h4>Wounds</h4>
            <p class=\\"wounds\\">30</p>
            <h4>Bindings</h4>
            <p class=\\"bindings\\">0</p>
            <h4>Shards</h4>
            <p class=\\"shards\\">18</p>
            <h4>Sidekicks</h4>
            <p class=\\"sidekicks\\">15</p>
            <h4>Ambitions</h4>
            <p class=\\"ambitions\\">30</p>
        "
`;

exports[`ui controller renders cleanup reset correctly 1`] = `
"
            <h2>Clean Up</h2>
            <h4>Bystanders</h4>
            <p class=\\"bystanders\\"></p>
            <h4>Wounds</h4>
            <p class=\\"wounds\\"></p>
            <h4>Bindings</h4>
            <p class=\\"bindings\\"></p>
            <h4>Shards</h4>
            <p class=\\"shards\\"></p>
            <h4>Sidekicks</h4>
            <p class=\\"sidekicks\\"></p>
            <h4>Ambitions</h4>
            <p class=\\"ambitions\\"></p>
        "
`;

exports[`ui controller renders form reset correctly 1`] = `
"
        
            <div class=\\"scheme-container\\">
                <h4 class=\\"text-danger\\">Scheme</h4>
                <p class=\\"text-danger scheme\\"></p>
            </div>
            
            <div class=\\"mastermind-container\\">
                <h4>Mastermind</h4>
                <p class=\\"mastermind\\"></p>
            </div>
            
            <div class=\\"villains-container\\">
                <h4 class=\\"text-danger\\">Villains</h4>
                <ul class=\\"villains row\\"></ul>
            </div>
            
            <div class=\\"heroes-container\\">
                <h4>Heroes</h4>
                <ul class=\\"heroes row\\"></ul>
            </div>
            
        "
`;

exports[`ui controller renders setup correctly 1`] = `
"
        
            <div class=\\"scheme-container\\">
                <h4 class=\\"text-danger\\">Scheme</h4>
                <p class=\\"text-danger scheme\\">
        <p class=\\"text-danger\\">Pull Reality into the Negative Zone</p>
        <img src=\\"../img/legendary-nick-fury.png\\">
    </p>
            </div>
            
            <div class=\\"mastermind-container\\">
                <h4>Mastermind</h4>
                <p class=\\"mastermind\\">
        <p class=\\"\\">Galactus</p>
        <img src=\\"../img/legendary-iron-man.png\\">
    </p>
            </div>
            
            <div class=\\"villains-container\\">
                <h4 class=\\"text-danger\\">Villains</h4>
                <ul class=\\"villains row\\">
            <div class=\\"col-sm\\">
                <p class=\\"image-label text-danger\\">Heralds of Galactus</p>
                <img src=\\"../img/legendary-captain-america.png\\">
            </div>   
        
            <div class=\\"col-sm\\">
                <p class=\\"image-label text-danger\\">Subterranea</p>
                <img src=\\"../img/legendary-captain-america.png\\">
            </div>   
        
            <div class=\\"col-sm\\">
                <p class=\\"image-label text-danger\\">Skrulls</p>
                <img src=\\"../img/legendary-captain-america.png\\">
            </div>   
        
            <div class=\\"col-sm\\">
                <p class=\\"image-label text-danger\\">Doombot Legion</p>
                <img src=\\"../img/legendary-captain-america.png\\">
            </div>   
        
            <div class=\\"col-sm\\">
                <p class=\\"image-label text-danger\\">Savage Land Mutates</p>
                <img src=\\"../img/legendary-captain-america.png\\">
            </div>   
        </ul>
            </div>
            
            <div class=\\"heroes-container\\">
                <h4>Heroes</h4>
                <ul class=\\"heroes row\\">
            <div class=\\"col-sm\\">
                <p class=\\"image-label\\">Mr. Fantastic</p>
                <img src=\\"../img/legendary-cyclops.png\\">
            </div>   
        
            <div class=\\"col-sm\\">
                <p class=\\"image-label\\">Invisible Woman</p>
                <img src=\\"../img/legendary-cyclops.png\\">
            </div>   
        
            <div class=\\"col-sm\\">
                <p class=\\"image-label\\">Human Torch</p>
                <img src=\\"../img/legendary-cyclops.png\\">
            </div>   
        
            <div class=\\"col-sm\\">
                <p class=\\"image-label\\">Thing</p>
                <img src=\\"../img/legendary-cyclops.png\\">
            </div>   
        
            <div class=\\"col-sm\\">
                <p class=\\"image-label\\">Spider-Man</p>
                <img src=\\"../img/legendary-cyclops.png\\">
            </div>   
        </ul>
            </div>
            
        "
`;
